File: Programmers/yerim/양과 늑대.cpp
// [92343] 양과 늑대
// dfs, 완전 탐색
#include <string>
#include <vector>
using namespace std;

vector<int> graph[17];
int ans = 1, s, w;

void dfs(int curNode, int w, int s, vector<int> nextNodes, vector<int> info) {
    if(info[curNode] == 0) s++;
    else w++;
    
    ans = max(ans, s);
    
    // 늑대 수 >= 양 수  -> 더 이상 탐색 진행하지 않음
    // 지금 당장은 양 수가 부족해서 탐색 진행하지 않는 것. 
    // nextNodes에 담겨있기 때문에 다른 곳에서 양을 모은 후 다시 탐색할 수 있음
    if(w >= s) return;
    
    for(int i = 0; i < nextNodes.size(); i++) {
        vector<int> next = nextNodes;
        next.erase(next.begin() + i);
        for(int j = 0; j < graph[nextNodes[i]].size(); j++) {
            next.push_back(graph[nextNodes[i]][j]);
        }
        dfs(nextNodes[i], w, s, next, info);
    }
}

int solution(vector<int> info, vector<vector<int>> edges) {
    for(int i = 0; i < edges.size(); i++) {
        graph[edges[i][0]].push_back(edges[i][1]);
    }
    
    vector<int> nextNodes;
    for(int i = 0; i < graph[0].size(); i++) {
        nextNodes.push_back(graph[0][i]);
    }
    dfs(0, 0, 0, nextNodes, info);
    return ans;
}

File: Programmers/chaeeun/양과 늑대.py
# (참고) 그래프. 백트래킹

def solution(info, edges):
    def nextNodes(v):
        temp = list()
        for e in edges:
            i, j = e # i는 부모노드, j는 자식노드
            if v == i: # 부모노드 번호 비교
                temp.append(j)
        return temp

    def dfs(sheep, wolf, current, path):
        ### 지금 노드 확인, 양 늑대 판별
        if info[current]:
            wolf += 1
        else:
            sheep += 1

        ### 늑대가 다 잡아먹음, 무시
        if sheep <= wolf:
            return 0

        ### 아니라면 임시 변수에 값 갱신
        maxSheep = sheep

        ### 서칭 시작
        for p in path:
            for n in nextNodes(p):
                if n not in path:
                    path.append(n)
                    ### 최대 양 판별
                    maxSheep = max(maxSheep, dfs(sheep, wolf, n, path))
                    path.pop()
        return maxSheep

    answer = dfs(0, 0, 0, [0])

    return answer
